internal/ls/utilities.go

```diff
@@ -711,6 +1006,512 @@ func nodeEndsWith(n *ast.Node, expectedLastToken ast.Kind, sourceFile *ast.Sourc
+		}
+	case ast.KindClassExpression:
+		// for class expressions, use the `class` keyword when the class is unnamed
+		return findChildOfKind(node, ast.KindClassKeyword, sourceFile)
+	case ast.KindFunctionExpression:
+		// for function expressions, use the `function` keyword when the function is unnamed
+		return findChildOfKind(node, ast.KindFunctionKeyword, sourceFile)
+	case ast.KindConstructor:
+		return node
+	}
+	return nil
+}
+
+func getAdjustedLocationForImportDeclaration(node *ast.ImportDeclaration, forRename bool) *ast.Node {
+	if node.ImportClause != nil {
+		if name := node.ImportClause.Name(); name != nil {
+			if node.ImportClause.AsImportClause().NamedBindings != nil {
+				// do not adjust if we have both a name and named bindings
+				return nil
+			}
+			// /**/import [|name|] from ...;
+			// import /**/type [|name|] from ...;
+			return node.ImportClause.Name()
+		}
+
+		// /**/import { [|name|] } from ...;
+		// /**/import { propertyName as [|name|] } from ...;
+		// /**/import * as [|name|] from ...;
+		// import /**/type { [|name|] } from ...;
+		// import /**/type { propertyName as [|name|] } from ...;
+		// import /**/type * as [|name|] from ...;
+		if namedBindings := node.ImportClause.AsImportClause().NamedBindings; namedBindings != nil {
+			if namedBindings.Kind == ast.KindNamedImports {
+				// do nothing if there is more than one binding
+				elements := namedBindings.AsNamedImports().Elements
+				if len(elements.Nodes) != 1 {
+					return nil
+				}
+				return elements.Nodes[0].Name()
+			} else if namedBindings.Kind == ast.KindNamespaceImport {
+				return namedBindings.Name()
+			}
+		}
+	}
+	if !forRename {
+		// /**/import "[|module|]";
+		// /**/import ... from "[|module|]";
+		// import /**/type ... from "[|module|]";
+		return node.ModuleSpecifier
+	}
+	return nil
+}
+
+func getAdjustedLocationForExportDeclaration(node *ast.ExportDeclaration, forRename bool) *ast.Node {
+	if node.ExportClause != nil {
+		// /**/export { [|name|] } ...
+		// /**/export { propertyName as [|name|] } ...
+		// /**/export * as [|name|] ...
+		// export /**/type { [|name|] } from ...
+		// export /**/type { propertyName as [|name|] } from ...
+		// export /**/type * as [|name|] ...
+		if node.ExportClause.Kind == ast.KindNamedExports {
+			// do nothing if there is more than one binding
+			elements := node.ExportClause.AsNamedExports().Elements
+			if len(elements.Nodes) != 1 {
+				return nil
+			}
+			return elements.Nodes[0].Name()
+		} else if node.ExportClause.Kind == ast.KindNamespaceExport {
+			return node.ExportClause.Name()
+		}
+	}
+	if !forRename {
+		// /**/export * from "[|module|]";
+		// export /**/type * from "[|module|]";
+		return node.ModuleSpecifier
+	}
+	return nil
+}
+
+func getMeaningFromLocation(node *ast.Node) ast.SemanticMeaning {
+	node = getAdjustedLocation(node, false /*forRename*/, nil)
```

Comment:
I don't know if this needs to handle JS-reparsed nodes, but if so, it should handle JSExportAssignment here and 

	KindJSTypeAliasDeclaration (type)
	KindCommonJSExport (value)
	KindJSImportDeclaration (all)

in getMeaningFromDeclaration